### DISCUSSION TITLE
Tracking only array length

### DISCUSSION THREAD

**Discussion by stellarhoof** [mobxjs/mobx#3552] [Q&A] [answered]
How come simply accessing the array length is tracked when the array stays the same size?

```javascript
import { action, autorun, observable } from "mobx"

let numbers = observable([1, 2, 3])
autorun(() => console.log(numbers.length))
action(() => (numbers[2] = 10))()
```

In the above example, the tracked function inside `autorun` runs twice.

Also, is there any way of tracking the array size only?

**Comment by urugator** [comment, marked as answer]
Atm observable array is backed by single `Atom`. We would need a seperate atom for `length`. Tbh dunno if it's intentional, legacy reasons or oversight. cc @mweststrate

> Also, is there any way of tracking the array size only?

No, but you can wrap it in extra `computed` to prevent the outer derivation from re-running.
```javascript
autorun(() => console.log(computed(() => numbers.length).get()))
```

**Reply by stellarhoof** [reply]
Thank you

**Comment by mweststrate** [comment]
Yes it is an intentional performance trade off, to avoid creating to many
subscriptions when looping / mapping over an array.

Note that a simpler way to achieve the above is reaction(() =>
array.length, length => { effect });

**Reply by urugator** [reply]
> to avoid creating to many subscriptions

Not disputing, just making sure we are on the same page here - the issue isn't about subscription per index, but an extra atom for `length` to differentiate length changes from value changes (of *any* index). So we are talking about maximum of two subscriptions instead of one per array.

**Reply by stellarhoof** [reply]
> `reaction(() => array.length, length => { effect })`

This was actually my first approach which has the downside of having to dispose of the reaction. I ended up going with @urugator's suggestion of `computed`. Here's the before/after:

Before:

```javascript
let useReactiveMemo = (fn) => {
  let [value, setValue] = React.useState(() => untracked(fn))
  React.useEffect(
    () =>
      reaction(fn, (next) =>
        setValue((prev) => (_.isEqual(prev, next) ? prev : next))
      ),
    []
  )
  return value
}
```

After:

```javascript
let useComputed = (fn) => _.head(React.useState(() => computed(fn, { equals: comparer.structural })))
```

**Reply by mweststrate** [reply]
> the issue isn't about subscription per index, but an extra atom for length to differentiate length changes from value changes (of any index)

@urugator sorry, missed that, the idea never occurred to me, but that should work!

**Reply by mweststrate** [reply]
> This was actually my first approach which has the downside of having to dispose of the reaction

@stellarhoof not sure if I parse the sentence correctly, but both `autorun` and `reaction` need disposing.

**Reply by stellarhoof** [reply]
The code I'm working on is `reaction` vs. `computed`. I just picked the `autorun` as an example.

But I'm glad you're open to tracking `length` as a different atom 👍

I can give this a shot if you'd like
